### PR TITLE
Outlier_detection: enforce max_ejection_percentage

### LINKED ISSUE #26877: Bug: Ejection algo removes more endpoints than max_ejection_percentage
*Title*: Ejected hosts should always be less than max_ejection_percentage

*Current behaviour*:
> Outlier detection ejects two pods when 3 total are running and max_ejection_percentage is 50%

*Expected behaviour*:
> max_ejection_percentage of 50% should only allow 1 ejection in case of 3 total pods.

[optional *Relevant Links*:]
> [Current algo checks](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier#ejection-algorithm) the limits first and ejection happens later -> https://github.com/envoyproxy/envoy/blob/54809a0530b26a33b21f77429be1b87772cca626/source/common/upstream/outlier_detection_impl.cc#L476-L483

Why not check the actual ejected_percentage that we will reach after this ejection? The current behaviour is a little counter-intuitive.

### PR DESCRIPTION
**Pull request by Pawan-Bishnoi**
Commit Message: Currently one host is always ejected regardless of max_ejection_percent.
Additional Description: The existing behaviour is counter intuitive. It does not respect max_ejection_percent (which should be the _true_ limit)

Risk Level: Medium
Testing: Test case exists
Docs Changes: todo
Release Notes: todo
Platform Specific Features:
Runtime guard: todo
Fixes #26877

### CODE REVIEW

File: source/common/upstream/outlier_detection_impl.cc [hunk 1]
```diff
@@ -480,8 +480,7 @@ void DetectorImpl::ejectHost(HostSharedPtr host,
   double ejected_percent = 100.0 * (ejections_active_helper_.value() + 1) / host_monitors_.size();
   // Note this is not currently checked per-priority level, so it is possible
   // for outlier detection to eject all hosts at any given priority level.
-  // Note: at-least one host is ejected, we ignore max ejection percentage when ejecting first host.
-  if ((ejections_active_helper_.value() == 0) || (ejected_percent <= max_ejection_percent)) {
+  if (ejected_percent <= max_ejection_percent) {
```
Comment (KBaichoo) [inline source/common/upstream/outlier_detection_impl.cc:483]: should be runtime guarded

File: test/common/upstream/outlier_detection_impl_test.cc [hunk 2]
```diff
@@ -483,8 +485,10 @@ TEST_F(OutlierDetectorImplTest, ConnectSuccessWithOptionalHTTP_OK) {
  * code.
  */
 TEST_F(OutlierDetectorImplTest, ExternalOriginEventsNonSplit) {
+  ON_CALL(runtime_.snapshot_, getInteger(MaxEjectionPercentRuntime, _)).WillByDefault(Return(60));
   EXPECT_CALL(cluster_.prioritySet(), addMemberUpdateCb(_));
   addHosts({"tcp://127.0.0.1:80"});
+  addHosts({"tcp://127.0.0.2:80"});
```
Comment (KBaichoo) [inline test/common/upstream/outlier_detection_impl_test.cc:491]: same host as above?

**Review by KBaichoo — COMMENTED** [review]
Thanks for working on this.

Can we also test (if not already there) that we won't eject host we have a single host and mep < 100? I think that was one of the original cases mentioned in https://github.com/envoyproxy/envoy/issues/26877#issuecomment-1562356124

File: test/common/upstream/outlier_detection_impl_test.cc (hunk 2 above)
Comment (Pawan-Bishnoi) [reply test/common/upstream/outlier_detection_impl_test.cc:491]: should I change the port?

Right now the IPs are different. Would `127.0.0.1:81` be better?

File: source/common/upstream/outlier_detection_impl.cc (hunk 1 above)
Comment (Pawan-Bishnoi) [reply source/common/upstream/outlier_detection_impl.cc]: done

**Review by KBaichoo — COMMENTED** [review]
Thanks for working on this.

/wait for CI

**Review by KBaichoo — APPROVED** [review]
Thank you for your contribution.

**Comment by Pawan-Bishnoi** [comment]
> Thank you for your contribution.

Thank you for patiently guiding me :smile:

I think I discovered a CI bug in the process. Will file one with details.

**Comment by fatedier** [comment]
@Pawan-Bishnoi Hi, I am using the OutlierDetection feature of Istio. Since the number of service replicas is dynamically changing, I hope that at least one replica will be ejected in case of a failure. We have set the max_ejection_percent to 10, but some services only have two or three Pods.

I have noticed a change in behavior in the new version of Istio where the previously working configuration does not take effect when a failure occurs. I have traced the relevant code changes and suspect they may be related to the modifications here, but I am not certain.

https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/outlier_detection.proto#:~:text=30000ms%20or%2030s.-,max_ejection_percent,-(UInt32Value)%20The

> The maximum % of an upstream cluster that can be ejected due to outlier detection. Defaults to 10% but will eject at least one host regardless of the value.

For example, when I have two pods, one of the pods is not healthy when it fails, and it only takes effect when I set max_ejection_percent to 50. But this is not the behavior I expected, I don't really want to set it to 50.

I hope you can provide some help.

**Comment by Pawan-Bishnoi** [comment]
if max_ejection_percent is 10% why do you want more than 10% pods to be ejected?

>  But this is not the behavior I expected, I don't really want to set it to 50.
why not?

**Comment by fatedier** [comment]
@Pawan-Bishnoi

In our scenario, this function is mainly aimed at reducing the impact of node failures on the service. Usually, only one or two nodes fail at the same time.

When there are few replicas, I want at least one fault instance to be ejected, rather than having this configuration not work at all. (Just as currently described in the Envoy documentation)

Traffic and the number of replicas are dynamically changing. When there are too many replicas, we do not want 50% of instances to be ejected, as this could potentially lead to more serious issues.

Will this change cause the OutlierDetection configuration of services with fewer than 10 replicas online to be completely ineffective? If I want to maintain the previous behavior as much as possible, how should I modify the configuration? I do not want to calculate the appropriate thresholds for each service precisely, especially when the number of replicas of services in the cluster changes dynamically.

**Comment by Pawan-Bishnoi** [comment]
@fatedier I can add that as an optional config.

This work was runtime guarded, you can use the runtime flag to continue using the old behaviour meanwhile: https://github.com/envoyproxy/envoy/pull/27909/commits/17deb412d0dab6cd5aa1e4dc9ff3e9ca50bae52a

**Comment by Pawan-Bishnoi** [comment]
This was removed since the default behaviour was really confusing and max_ejection_percent did not really mean what the name says.

But your usecase seems fair and it can be enabled intentionally via some config variable. I will how this PR for now (https://github.com/envoyproxy/envoy/pull/33657) and work on converting the runtime var to a flag instead.

**Pull request merged**